[Driver] Add support for crtbegin.o, crtend.o and libgloss lib to BareMetal toolchain object

quic-garvgupt opened the pull request:
This patch conditionalise the addition of crt{begin,end}.o object files along
with addition of -lgloss lib based on whether libc selected is newlib or llvm
libc. Since there is no way a user can specify which libc it wants to link
against, currently passing valid GCCInstallation to driver will select newlib
otherwise it will default to llvm libc.

Moreover, this patch makes gnuld the default linker for baremetal toolchain
object. User need to pass `-fuse-ld=lld` explicitly to driver to select lld

This is the 2nd patch in the series of patches of merging RISCVToolchain into
BareMetal toolchain object.

RFC:
https://discourse.llvm.org/t/merging-riscvtoolchain-and-baremetal-toolchains/75524

quic-garvgupt commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121830?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#121831** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121831?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#121830** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121830?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121830?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#121829** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121829?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

MaskRay commented on clang/test/Driver/arm-toolchain.c:
```diff
@@ -59,4 +97,39 @@
+// ARM-BAREMETAL-UNWINDLIB: "{{.*}}crt0.o"
+// ARM-BAREMETAL-UNWINDLIB: "{{.*}}clang_rt.crtbegin.o"
+// ARM-BAREMETAL-UNWINDLIB: "--start-group" "{{.*}}libclang_rt.builtins.a" "--as-needed" "-lunwind" "--no-as-needed" "-lc" "-lgloss" "--end-group"
+// ARM-BAREMETAL-UNWINDLIB: "{{.*}}clang_rt.crtend.o"
```
all the modified test files don't end with `\n` (see vim `:h eol`)

quic-garvgupt replied:
Resolved

quic-garvgupt commented:
Gentle Ping!

topperc commented on clang/lib/Driver/ToolChains/BareMetal.h:
```diff
@@ -36,6 +36,7 @@ class LLVM_LIBRARY_VISIBILITY BareMetal : public Generic_ELF {
   Tool *buildStaticLibTool() const override;
 
 public:
+  virtual bool isUsingLD() const { return UseLD || GCCInstallation.isValid(); }
```
Does this need to be virtual?

quic-garvgupt replied:
There was a request in the original PR that there is some downstream toolchain that inherits from Baremetal toolchain and defaults to LLD. Therefore I made this virtual so that it cab be overridden however, in the recent patchset I have removed it since there is no need for it in community and can easily be maintained downstream.

quic-garvgupt commented:
Gentle Ping again!

quic-garvgupt commented:
Hi @petrhosek, following up on our last RISC-V embedded sync-up, can you please review this patch? Thanks

MaskRay commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -154,6 +154,8 @@ BareMetal::BareMetal(const Driver &D, const llvm::Triple &Triple,
   GCCInstallation.init(Triple, Args);
   SysRoot = computeSysRoot();
+  UseLD =
+      Args.getLastArgValue(options::OPT_fuse_ld_EQ).equals_insensitive("ld");
```
should not use equals_insensitive. It's only for Windows.

quic-garvgupt replied:
Removed the assignment statement itself since it became redundant

MaskRay commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -331,6 +333,32 @@ BareMetal::OrderedMultilibs BareMetal::getOrderedMultilibs() const {
+}
+
+const char *BareMetal::getDefaultLinker() const {
+  if (isUsingLD())
```
this ld/ld.lld selection feels very odd. Ideally we just stick with one `ld`. If a user prefers `ld.lld`, customize this preference in Clang configuration file.
The shared C++ code trying to appease some sets of users could easily end up with something with unprecdictable behavior, which offers worse user experience for the other set of users.

petrhosek replied:
I agree, I understand the desire to match the existing behavior, but in this case I think it's counterproductive, users should the existing options like `-fuse-ld` and ` to choose a default linker `CLANG_DEFAULT_LINKER`.

quic-garvgupt replied:
Fixed. ld will be the default Linker

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -524,12 +552,21 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   const llvm::Triple &Triple = getToolChain().getEffectiveTriple();
 
-  AddLinkerInputs(TC, Inputs, Args, CmdArgs, JA);
+  if (!D.SysRoot.empty())
+    CmdArgs.push_back(Args.MakeArgString("--sysroot=" + D.SysRoot));
```
Can you move this to a separate PR?

quic-garvgupt replied:
Done
[Driver] Forward sysroot from Driver to linker in BareMetal ToolChain Object https://github.com/llvm/llvm-project/pull/132808

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -524,12 +552,21 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+  if (Triple.isRISCV()) {
+    if (Args.hasArg(options::OPT_mno_relax))
+      CmdArgs.push_back("--no-relax");
+    if (TC.isUsingLD()) {
+      CmdArgs.push_back("-m");
+      CmdArgs.push_back(Arch == llvm::Triple::riscv64 ? "elf64lriscv"
+                                                      : "elf32lriscv");
+    }
```
This option is supported by both LLD and GNU linkers as well so there's no need to conditionalize this on the linker selection.

quic-garvgupt replied:
Fixed

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -524,12 +552,21 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
 
-  if (TC.getTriple().isRISCV() && Args.hasArg(options::OPT_mno_relax))
-    CmdArgs.push_back("--no-relax");
+  if (Triple.isRISCV()) {
+    if (Args.hasArg(options::OPT_mno_relax))
+      CmdArgs.push_back("--no-relax");
+    if (TC.isUsingLD()) {
+      CmdArgs.push_back("-m");
+      CmdArgs.push_back(Arch == llvm::Triple::riscv64 ? "elf64lriscv"
+                                                      : "elf32lriscv");
+    }
+    CmdArgs.push_back("-X");
+  }
```
Can you move this to a separate PR?

quic-garvgupt replied:
Done
[RISCV] Integrate RISCV target in baremetal toolchain object and deprecate RISCVToolchain object https://github.com/llvm/llvm-project/pull/121831

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -565,26 +637,16 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     AddRunTimeLibs(TC, D, CmdArgs, Args);
-
     CmdArgs.push_back("-lc");
+    if (TC.isUsingLD())
+      CmdArgs.push_back("-lgloss");
```
This is strange since using libgloss has nothing to do with the linker selection, rather it's related to the libc choice since libgloss is a part of Newlib. It looks like `RISCVToolchain` assumes Newlib but we can't make that assumption for `BareMetal` since it's also being used with other libs like LLVM libc.

We might need to consider introduce a new option for libc selection into Clang, but for your change it might be OK to use `GCCInstallation.isValid()` as a short-term workaround to match the existing logic.

petrhosek replied:
I proposed the libc selection as a topic for the next LLVM libc monthly meeting on March 13.

quic-garvgupt replied:
Fixed

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -540,19 +577,54 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   for (const auto &LibPath : TC.getLibraryPaths())
     CmdArgs.push_back(Args.MakeArgString(llvm::Twine("-L", LibPath)));
 
+  if (D.isUsingLTO()) {
+    assert(!Inputs.empty() && "Must have at least one input.");
+    // Find the first filename InputInfo object.
+    auto Input = llvm::find_if(
+        Inputs, [](const InputInfo &II) -> bool { return II.isFilename(); });
+    if (Input == Inputs.end())
+      // For a very rare case, all of the inputs to the linker are
+      // InputArg. If that happens, just use the first InputInfo.
+      Input = Inputs.begin();
+
+    addLTOOptions(TC, Args, CmdArgs, Output, *Input,
+                  D.getLTOMode() == LTOK_Thin);
+  }
```
Can you move this to a separate PR?

quic-garvgupt replied:
Done
[Driver] Fix link order of BareMetal toolchain object https://github.com/llvm/llvm-project/pull/132806

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -540,19 +577,54 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
 
-  Args.addAllArgs(CmdArgs, {options::OPT_L, options::OPT_T_Group,
-                            options::OPT_s, options::OPT_t, options::OPT_r});
+  Args.addAllArgs(CmdArgs,
+                  {options::OPT_L, options::OPT_u, options::OPT_T_Group,
+                   options::OPT_s, options::OPT_t, options::OPT_r});
```
Can you move this to a separate PR?

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -540,19 +577,54 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+  if (WantCRTs) {
+    if (!Args.hasArg(options::OPT_r))
+      CmdArgs.push_back(Args.MakeArgString(TC.GetFilePath("crt0.o")));
+    if (TC.isUsingLD()) {
+      auto RuntimeLib = TC.GetRuntimeLibType(Args);
+      if (RuntimeLib == ToolChain::RLT_Libgcc) {
+        crtbegin = "crtbegin.o";
+        crtend = "crtend.o";
+      } else {
+        assert(RuntimeLib == ToolChain::RLT_CompilerRT);
+        crtbegin =
+            TC.getCompilerRTArgString(Args, "crtbegin", ToolChain::FT_Object);
+        crtend =
+            TC.getCompilerRTArgString(Args, "crtend", ToolChain::FT_Object);
+      }
+      CmdArgs.push_back(Args.MakeArgString(TC.GetFilePath(crtbegin)));
+    }
```
Whether we're using `crtbegin.o` and `crtend.o` is not related to the linker selection, rather it depends on the libc and how it handles initialization. It looks like RISCVToolchain assumes Newlib but we can't make that assumption for BareMetal since it's also being used with other libs like LLVM libc.

We might need to consider introduce a new option for libc selection into Clang, but for your change it might be OK to use GCCInstallation.isValid() as a short-term workaround to match the existing logic.

quic-garvgupt replied:
Fixed

petrhosek commented:
I asked for various parts of the linker setup to be moved into a separate PR. I understand that each of those changes would cause some churn in tests, but I think that's desirable so we understand the effect of each of those changes in isolation.

Commit:
[Driver] Add support for crtbegin.o, crtend.o and libgloss lib to BareMetal
toolchain object

This patch conditionalise the addition of crt{begin,end}.o object files along
with addition of -lgloss lib based on whether libc selected is newlib or llvm
libc. Since there is no way a user can specify which libc it wants to link
against, currently passing valid GCCInstallation to driver will select newlib
otherwise it will default to llvm libc.

Moreover, this patch makes gnuld the default linker for baremetal toolchain
object. User need to pass `-fuse-ld=lld` explicitly to driver to select lld

This is the 2nd patch in the series of patches of merging RISCVToolchain into
BareMetal toolchain object.

RFC:
https://discourse.llvm.org/t/merging-riscvtoolchain-and-baremetal-toolchains/75524

Change-Id: Ie06dc976c306cf04ec2733bbb2d271c57d201f86

quic-garvgupt commented:
> I asked for various parts of the linker setup to be moved into a separate PR. I understand that each of those changes would cause some churn in tests, but I think that's desirable so we understand the effect of each of those changes in isolation.

I have split the PR as requested. However, this PR was pushed a little over two months ago, and such requests would have been greatly appreciated earlier in the review process. Implementing significant changes at later stages can be challenging. I hope that the pace of the reviews will now be more consistent. Thank you very much for all your feedback and comments.

quic-garvgupt commented:
Hi @petrhosek , if this patch looks good after splitting, please approve the PR. I would like to start merging the patches in this series. Thanks!

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -545,9 +545,27 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+  bool WantCRTs =
+      !Args.hasArg(options::OPT_nostdlib, options::OPT_nostartfiles);
+
+  const char *crtbegin, *crtend;
```
This is probably more consistent with LLVM style:
```suggestion
  const char *CRTBegin, *CRTEnd;
```

quic-garvgupt replied:
Done

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -545,9 +545,27 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+    if (!Args.hasArg(options::OPT_r))
+      CmdArgs.push_back(Args.MakeArgString(TC.GetFilePath("crt0.o")));
+    if (TC.hasValidGCCInstallation() || hasGCCToolChainAlongSideClang(D)) {
+      auto RuntimeLib = TC.GetRuntimeLibType(Args);
+      if (RuntimeLib == ToolChain::RLT_Libgcc) {
+        crtbegin = "crtbegin.o";
+        crtend = "crtend.o";
+      } else {
+        assert(RuntimeLib == ToolChain::RLT_CompilerRT);
+        crtbegin =
+            TC.getCompilerRTArgString(Args, "crtbegin", ToolChain::FT_Object);
+        crtend =
+            TC.getCompilerRTArgString(Args, "crtend", ToolChain::FT_Object);
```
I think a `switch` is a better fit here:
```suggestion
      auto RuntimeLib = TC.GetRuntimeLibType(Args);
      switch (TC.GetRuntimeLibType(Args)) {
      case ToolChain::RLT_Libgcc) {
        crtbegin = "crtbegin.o";
        crtend = "crtend.o";
        break;
      }
      case ToolChain::RLT_CompilerRT: {
        crtbegin =
            TC.getCompilerRTArgString(Args, "crtbegin", ToolChain::FT_Object);
        crtend =
            TC.getCompilerRTArgString(Args, "crtend", ToolChain::FT_Object);
        break;
      }
```

quic-garvgupt replied:
Done

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -545,9 +545,27 @@ void baremetal::Linker::ConstructJob(Compilation &C, const JobAction &JA,
-  if (!Args.hasArg(options::OPT_nostdlib, options::OPT_nostartfiles,
-                   options::OPT_r)) {
-    CmdArgs.push_back(Args.MakeArgString(TC.GetFilePath("crt0.o")));
+  bool WantCRTs =
```
This is just a suggestion, but I think `NeedsCRT` would be more accurate.
```suggestion
  bool NeedCRT =
```

quic-garvgupt replied:
Done

quic-garvgupt commented:
Ping!

MaskRay commented on clang/test/Driver/arm-toolchain.c:
```diff
@@ -1,5 +1,12 @@
 // UNSUPPORTED: system-windows
 
+// Test interaction with -fuse-ld=lld
```
You need to ensure that there is an `ld.lld` file with `-B%S/Inputs/lld`

quic-garvgupt replied:
Done

Kewen12 commented:
Hi @quic-garvgupt, we are seeing llvm check errors with this PR that breaks our buildbots. Could you please take a look? Thanks!

> FAIL: Clang::aarch64-toolchain.c
FAIL: Clang::arm-toolchain.c

bot: https://lab.llvm.org/buildbot/#/builders/10/builds/7529

lenary commented:
@kewen12 I think I can see how to fix these tests, I'm just confused because these tests seem to end up requesting both compiler-rt and libgcc. Is that bot configured with a different default unwindlib to normal or something?

lenary commented:
My plan is to modify the tests to accept that there might be unwindlib arguments between `clang_rt.builtins.a` and `-lc`, which is where the breakage is right now.
